helix-gateway/src/main/java/org/apache/helix/gateway/util/GatewayCurrentStateCache.java

```diff
@@ -0,0 +1,144 @@
+      Map<String, String> newCurrentState = entry.getValue();
+      Map<String, String> oldCurrentState = _currentStateMap.get(instance);
+      if (oldCurrentState == null) {
+        diff.put(instance, newCurrentState);
+        continue;
+      }
+      if (!oldCurrentState.equals(newCurrentState)) {
+        for (String shard : newCurrentState.keySet()) {
+          if (oldCurrentState.get(shard) == null || !oldCurrentState.get(shard).equals(newCurrentState.get(shard))) {
+            diff.computeIfAbsent(instance, k -> new HashMap<>()).put(shard, newCurrentState.get(shard));
+          }
+        }
+      }
+    }
+    _currentStateMap = newCurrentStateMap;
+    return diff;
+  }
+
+  public void updateCacheWithCurrentStateDiff(Map<String, Map<String, String>> currentStateDiff) {
+    updateShardStateMapWithDiff(currentStateDiff, _currentStateMap);
+  }
+
+  /**
+   * Udate the target state with the changed target state maps.
+   * All existing target states remains the same
+   * @param targetStateChangeMap
+   */
+  public void updateTargetStateWithDiff(Map<String, Map<String, String>> targetStateChangeMap) {
+    synchronized (_targetStateLock) {
+      _targetStateChanged = updateShardStateMapWithDiff(targetStateChangeMap, _targetStateMap);
+    }
+  }
+
+  public Map<String, Map<String, String>> getTargetStateMapIfChanged() {
+    synchronized (_targetStateLock) {
+      if (_targetStateChanged) {
+        _targetStateChanged = false;
+        return _targetStateMap;
+      }
+      return null;
+    }
+  }
+
+  public String serializeTargetAssignments() {
+    ObjectMapper mapper = new ObjectMapper();
+    ObjectNode root = mapper.createObjectNode();
+    synchronized (_targetStateLock) {
+      for (Map.Entry<String, Map<String, String>> entry : _targetStateMap.entrySet()) {
+        String instance = entry.getKey();
+        Map<String, String> assignments = entry.getValue();
+        ObjectNode instanceNode = mapper.createObjectNode();
+        for (Map.Entry<String, String> assignment : assignments.entrySet()) {
+          instanceNode.put(assignment.getKey(), assignment.getValue());
+        }
+        root.set(instance, instanceNode);
+      }
+    }
+    return root.toString();
+  }
+
+  private boolean updateShardStateMapWithDiff(Map<String, Map<String, String>> diffMap,
+      Map<String, Map<String, String>> currentMap) {
+    if (diffMap == null || diffMap.isEmpty()) {
+      return false;
+    }
+    for (Map.Entry<String, Map<String, String>> entry : diffMap.entrySet()) {
+      String instance = entry.getKey();
+      Map<String, String> currentState = entry.getValue();
+      if (currentMap.get(instance) == null) {
+        currentMap.put(instance, currentState);
+      } else {
+        currentMap.get(instance).entrySet().stream().forEach(e -> {
+          if (currentState.get(e.getKey()) != null) {
+            e.setValue(currentState.get(e.getKey()));
+          }
+        });
```

Comment:
Why update the entry of the diff map? That's tmp object.

Reply:
We are updating the `currentMap`, with diff map value.